### ISSUE TITLE
I'm on the Authors list twice

### ISSUE THREAD

**Issue by illegitimate-egg** [o2sh/onefetch#923]
### Duplicates

- [X] I have searched the existing issues

### Current behavior 😯

Onefetch reports two Authors, when there is only one. Both authors are the same person.

### Expected behavior 🤔

Onefetch reports one Author.

### Steps to reproduce 🕹

1. git clone https://github.com/illegitimate-egg/ncurses-prime
2. onefetch

### Additional context/Screenshots 🔦

![image](https://user-images.githubusercontent.com/63966843/211193231-de73dfac-7c11-4e8a-bc23-74960e03c4c0.png)

**Comment by o2sh** [comment]
Hi @illegitimate-egg,

All your commits were not made withe the same email address:

![image](https://user-images.githubusercontent.com/13710835/211197405-732de235-f2be-4885-8131-d30746b72276.png)

To fix it you need to create a `.mailmap` file a the root of your repo (more info about [mailmap](https://git-scm.com/docs/gitmailmap)) where you can specify the correct mapping.

With this mapping:
![image](https://user-images.githubusercontent.com/13710835/211197493-3ae8db57-66c3-45cf-8624-a3f1e114441c.png)

You'll get this:
![image](https://user-images.githubusercontent.com/13710835/211197507-f1b4d1f5-4a33-4db0-a99a-772a24039492.png)

**Comment by illegitimate-egg** [comment]
Oh, thanks!

**Comment by spenserblack** [comment]
Do you think it's worth adding this to an FAQ? :thinking:

**Comment by illegitimate-egg** [comment]
> Do you think it's worth adding this to an FAQ? 🤔

It's more of a problem with me being inexperienced with git

**Comment by spenserblack** [comment]
To be fair, I think mailmap isn't a well-known feature :slightly_smiling_face:

**Comment by o2sh** [comment]
I've added an FAQ section to the wiki with an entry pointing to this issue: https://github.com/o2sh/onefetch/wiki/frequently-asked-questions

**Comment by o2sh** [comment]
>Do you think it's worth adding this to an FAQ? 🤔

![image](https://user-images.githubusercontent.com/13710835/211225315-478b3c78-097b-44ad-baa6-8801ab0b26e4.png)

Hopefully, people should stumble upon this issue before creating a duplicate.

**Comment by spenserblack** [comment]
Oh yeah, good point :laughing:

**Comment by beutlich** [comment]
Even with having a .mailmap file available I see duplicatd authors. Can anyone confirm with [this repo](https://github.com/modelica/ModelicaStandardLibrary)?

**Comment by spenserblack** [comment]
@beutlich Looks like your `.mailmap` lists every author and their email. The two most common use cases for mailmaps IMO is to fix the name mapped to an email, and provide a secondary email (and name) mapped to the author's primary info. For example, with this change alone, you're listed only once, since these are your two most frequently used emails (and also combine your two records if you use `git shortlog --email`):
```diff
diff --git a/.mailmap b/.mailmap
index 864f8ad46..5b7291e7b 100644
--- a/.mailmap
+++ b/.mailmap
@@ -114,8 +114,7 @@ Stephan Ziegler <stephan.ziegler@3ds.com>
 Sven Erik Mattson <svenerik@7ce873d0-865f-4ce7-a662-4bb36ea78beb>
 Thomas Beutlich <beutlich@7ce873d0-865f-4ce7-a662-4bb36ea78beb>
 Thomas Beutlich <beutlich@users.noreply.github.com>
-Thomas Beutlich <modelica@tbeu.de>
-Thomas Beutlich <thomas.beutlich@esi-group.com>
+Thomas Beutlich <modelica@tbeu.de> <thomas.beutlich@esi-group.com>
 Thomas Bödrich <ThomasBoedrich@7ce873d0-865f-4ce7-a662-4bb36ea78beb>
 Volker Waurich <12234403+vwaurich@users.noreply.github.com>
 carlj-w <41286817+carlj-w@users.noreply.github.com>
```

Also, this would work:

```diff
diff --git a/.mailmap b/.mailmap
index 864f8ad46..aede18718 100644
--- a/.mailmap
+++ b/.mailmap
@@ -112,10 +112,9 @@ Stefan Wischhusen <wischhusen@7ce873d0-865f-4ce7-a662-4bb36ea78beb>
 Stefan Wischhusen <wischhusen@xrg-simulation.de>
 Stephan Ziegler <stephan.ziegler@3ds.com>
 Sven Erik Mattson <svenerik@7ce873d0-865f-4ce7-a662-4bb36ea78beb>
-Thomas Beutlich <beutlich@7ce873d0-865f-4ce7-a662-4bb36ea78beb>
-Thomas Beutlich <beutlich@users.noreply.github.com>
-Thomas Beutlich <modelica@tbeu.de>
-Thomas Beutlich <thomas.beutlich@esi-group.com>
+<modelica@tbeu.de> <beutlich@7ce873d0-865f-4ce7-a662-4bb36ea78beb>
+<modelica@tbeu.de> <thomas.beutlich@esi-group.com>
+<modelica@tbeu.de> <beutlich@users.noreply.github.com>
 Thomas Bödrich <ThomasBoedrich@7ce873d0-865f-4ce7-a662-4bb36ea78beb>
 Volker Waurich <12234403+vwaurich@users.noreply.github.com>
 carlj-w <41286817+carlj-w@users.noreply.github.com>
```

More information here: https://git-scm.com/docs/gitmailmap

**Comment by beutlich** [comment]
Thanks for the fast reply @spenserblack.

I generated that .mailmap file with line-wise author listing based on https://kevin.burke.dev/kevin/easy-maintenance-of-your-authors-file/ and it worked out in git or TortoiseGit.

Is there any chance that onefetch could also consider these line-wise author mapping?

**Comment by spenserblack** [comment]
See https://github.com/o2sh/onefetch/issues/447 (and the linked PRs/issues in that discussion) for some more context, but I think we're pretty much set on sticking with email as the identifier.

I don't think that script was really intended to generate a `.mailmap` file -- quickly skimming that page seems to imply it was used to make a human-readable credits. Since your `.mailmap` was generated from a git log, I'm not really sure what it's actually accomplishing TBH. By "working out" do you mean "not breaking"? In fact, by updating your `.mailmap` to map all emails to the same person, I'd say it would *fix* [the linked script](https://kevin.burke.dev/kevin/easy-maintenance-of-your-authors-file/) by combining duplicates into one user (I tested this out on our own repository, where we have both same name + different email and same email + different name issues in our log :sweat_smile:).

If you [search for `.mailmap` files](https://github.com/search?q=path%3A%2F(%5E%7C%5C%2F)%5C.mailmap%2F&type=code) you should see that `Author Name <primary@example.com> <secondary@example.com>` is a popular format. Besides onefetch, I'd say it's a pretty good standard to have a `.mailmap`. It will also fix your own `git log` by providing a canonical email address/name in the author information.

**Comment by beutlich** [comment]
Thank your for your fast feedback.

> By "working out" do you mean "not breaking"?

Sorry for not being excat enough. The given [.mailmap](https://github.com/modelica/ModelicaStandardLibrary/blob/master/.mailmap) file with line-wise authors is correctly taken into account by both

* git command `git shortlog -sn --since="25 January 2014"`
* TortoiseGit Statistics dialog if set
  ![grafik](https://github.com/o2sh/onefetch/assets/14896695/6b33220e-a50a-447b-84c2-ec17cbb81ed2)
  
Also, https://git-scm.com/docs/gitmailmap mentions

```
Joe R. Developer <joe@example.com>
Jane Doe <jane@example.com>
Jane Doe <jane@desktop.(none)>
```

as the corrected example to fix the names.

Given these points, my expectation was that this kind of formatting is a valid format for .mailmap (even it its original intent of generation was some AUTHORS file).

**Comment by spenserblack** [comment]
> Also, https://git-scm.com/docs/gitmailmap mentions
> 
> ```
> Joe R. Developer <joe@example.com>
> Jane Doe <jane@example.com>
> Jane Doe <jane@desktop.(none)>
> ```
> 
> as the corrected example to fix the names.

Yes, which I interpret as the email being the person's identifier, and the name being a label. I.e. "this email has this name," not "this name has this email." Do you notice any change at all when you *uncheck* "use .mailmap"? Because, since your `.mailmap` is just a repetition of what is in the original commit information, I don't expect it would have any noticeable effect.

You might want to see [TortoiseGit's own `.mailmap`](https://github.com/TortoiseGit/TortoiseGit/blob/508dae4de515c55ad5c143a906af695be06bbf8f/.mailmap) :wink:

You are correct that `git shortlog -sn` will combine the records, but `git shortlog -sne` will separate them (unless you fix the `.mailmap` like the patches from https://github.com/o2sh/onefetch/issues/923#issuecomment-1910247681 demonstrate). We consider email to be the "more unique" and therefore more reliable identifier between between the committer's name and their email. Note that some popular tools, such as GitHub itself, use the email as the identifier, not the name. See https://github.com/spenserblack/test-repo/commit/7d50c70d77ba5b640a1d2b045cdabcaff754df79, https://github.com/spenserblack/test-repo/commit/b83afd5b8958cc899d0b5fd81d3deda111b6d006, and https://github.com/spenserblack/test-repo/commit/037f07e8a50cde350428fd7fd08ed97cc145d240, which are commits with a different author name, different author email, and committer email *and* author email each.

If your statement is "onefetch is not using the `.mailmap` correctly," currently I'm still not convinced that your `.mailmap` is valid (well, it's syntactically valid, but repeating all commit names and emails as proper names and emails is not that useful AFAIK). It's effectively stating that each record *is* a different person, since each line is only giving their proper name and email, not mapping their proper name and email to commit names and emails.

However, if your statement is "onefetch should combine matching author names into one record," then I believe there is some potential merit in that, given the low chances of two people having the same name in small to medium-sized projects, and I'd recommend raising a new issue so we could discuss that specifically. Even if we've already discussed and switched away from that, it could be good to at least reconsider.

**Comment by beutlich** [comment]
Thanks for the detailed reply and your offers. I now adapted the corresponding .mailmap file to the expected format. Unfortunately this was manual work.

**Comment by spenserblack** [comment]
:thinking: It's never pleasant to do manual work :sweat_smile:
For the most accuracy I think it's unavoidable -- a mailmap provides information the machine can't know with certainty: name changes, new emails, etc.

However, if you've gotten your mailmap into a good initial state now, it shouldn't be too difficult to write a script to detect any names or emails in the log that aren't in the mailmap. You could run this script in a GitHub workflow so that you're made aware when a new author is added (this can be triggered by a PR to catch it ahead of time or on push to the main branch to handle squash merges).

Personally, I've never bothered to add *every* author to a mailmap. Just the ones where I *know* the same person used two or more committer names/emails.

**Comment by beutlich** [comment]
Hm, as a side-effect of the .mailmap format change GitHub no longer displays the correct author mapping in the [commit history](https://github.com/modelica/ModelicaStandardLibrary/commits/master/).

![grafik](https://github.com/o2sh/onefetch/assets/14896695/cf19bfe4-409b-4979-946d-2700043d804b)

I guess I need to revert that change.
